server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -407,12 +475,27 @@ public String executor() {
     }
 
     public void sendValidateJoinRequest(DiscoveryNode node, ClusterState state, ActionListener<TransportResponse.Empty> listener) {
-        transportService.sendRequest(
-            node,
-            VALIDATE_JOIN_ACTION_NAME,
-            new ValidateJoinRequest(state),
-            new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
-        );
+        try {
+            BytesReference bytes;
+            if (clusterStateRefreshInterval == 0) {
+                bytes = CompressionHelper.serializeClusterState(state, node, true);
+            } else {
+                bytes = serializedStates.computeIfAbsent(
+                    node.getVersion(),
```

Comment:
I asked the question [here](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1551713787) as well, but can this be a composite key with `state.version()` so that we always recompute the state if a newer state exists and never send stale data? If so, we could probably get rid of the configurable interval setting, and possibly replace the cache object with simpler memoization.

Reply:
Responded [here](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1552738053) well. Just going with state-version based caching won't give enough benefit for customers. At least for the scenario we're trying to solve, evicting cache with every node-join update would be non-deterministic for the benefits. We may see some improvements sometimes but not all the time. 

A mixed approach of both state-version based caching and time based would be more helpful. We can keep the default behaviour of no caching in 2.x but change it for 3.0 with 30s cache to make sure customers see significant improvement during sudden multi node joins.

Let's close the approach on https://github.com/opensearch-project/OpenSearch/issues/5100 conversation, based on that I'll push next commit.

Thanks @andrross for all the useful comments 👍🏼

Reply:
@amkhar It isn't very clear as to what are the trade-offs of caching? Can we run into nodes with incompatible cluster state join the cluster? Whats the rationale behind 30s meaning what do we lose if we make this 30 mins? I would second @andrross's recommendation on memoization is the idea is that diffs would be minimal but frequent to invalidate caches on cluster manager.
IMO the settings makes me feel we are building a semantic that can be misused, can leak to breakages of cluster state guarantees with consistency in future.

Reply:
Thanks @Bukhtawar for review.
It would be easy to close the approach on https://github.com/opensearch-project/OpenSearch/issues/5100 conversation as it has all the background and experiment results which will answer some of the questions.

>Can we run into nodes with incompatible cluster state join the cluster?

Cluster state is sent by Cluster Manager, so if we send cluster state of 10s ago, data node is supposed to run validations on the index metadata of incoming cluster state with its node version to make sure correct version of data node will send a success in return of that join validation request(from cluster manager). So, node will run validations on that data of 10s ago. Now Cluster Manager will anyway run those validations as part of node-join task through TaskBatcher and fail the join if any validation fails. Only thing which we need to keep in mind is that, if any of the discovery plugin has some extra validation, that will work on 10s older state. Today, **we don't have any validators from ec2, azure or gce**.

>Whats the rationale behind 30s meaning what do we lose if we make this 30 mins?

Index created during this interval(30s or 30min) won't be part of the index metadata of cluster state which is being sent for validation to data nodes.

> I would second @andrross's recommendation on memoization is the idea is that diffs would be minimal but frequent to invalidate caches on cluster manager.

Sure, I'm running some tests with just state-version based caching where we invalidate the cache if cluster state is updated. I'll share the results so we can collaboratively take a decision.

Reply:
>Whats the rationale behind 30s

I choose this interval after running experiments on other values like 10s also. So most benefit was in 30s and it takes around 1 minute for 100 nodes to join the cluster back so whatever best caching we can provide in that interval is going to reflect in allocation gains.

Reply:
Update : using state's version based caching only after seeing results https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1557087555